Print heredocs consistently with lineSuffix(newline+contents+ending)

jbielick opened the pull request:
## Purpose

Instead of ad-hoc code in various places plucking out heredoc `beging`, `contents`, and `ending` in order to print it more intelligently in certain cases, this changes the way heredocs are printed altogether so they can simply be printed in any context and still work correctly. This prevents heredocs appearing in various parts of the syntax tree from _absorbing_ other lines of code because they break after the heredoc literal and before the heredoc contents because they were manually plucked out the AST and inserted in the Doc contextually. 

fixes prettier/plugin-ruby#624
fixes prettier/plugin-ruby#586
fixes prettier/plugin-ruby#316
fixes prettier/plugin-ruby#301

## Approach

I used `lineSuffix`.

The heredoc is 3 components, `beging` (the token), `contents` (the initial content and ending newline), and `ending` (the token ending).

```ruby
<<-TEXT
stuff stuff
stuff
TEXT
```

`beging`=`<<-TEXT`
`contents`=`stuff stuff\nstuff\n`
`ending`=`TEXT`

#### lineSuffix

[There exists a command in prettier's set of builders called `lineSuffix`](https://github.com/prettier/prettier/blob/master/commands.md#linesuffix). It's listed as used for implementing trailing comments. The doc contents you wrap in `lineSuffix()` are appended to the end of the line during printing. This seemed appropriate for this use case because of the following facts about heredocs:

1. A heredoc literal beging can appear anywhere and can be treated like a string literal. things can be chained to it, it may need a trailing comma, etc).

Conclusion: When printing a heredoc, more code should be able to print after the heredoc literal before the next line.

2. Everything after the first line break following the heredoc beging up until the ending **is part of the heredoc contents. period.**

Conclusion: A line break must _always_ appear on same line as the beging and the contents and ending must follow. Code can continue on the line after the ending.

Conclusion: Code chained or in the same expression as the heredoc beging might fit on the same line as the heredoc beging or be may need to be broken and continue on the newline following the ending.

3. Using `lineSuffix` when printing heredocs allows us to guarantee an append to the end of that line where we can insert a literalline (non-negotiable line break), the contents, and the ending. Code continues printing after this ending newline.

Conclusion: `lineSuffix`'s behavior matches the behavior of heredocs in general and would allow a heredoc to be printed correctly in any context. The heredoc print would look like this:

```js
concat([
  beging,
  lineSuffix(concat([literalline, concat(parts), ending]))
]);
```

In words: since a heredoc beging appears "here", print the beging "here", but guarantee the following contents will appear at the end of the "here" line: `linebreak, contents, ending`. 

#### Result:

Thus, the end of line break and next few lines are guaranteed to be the heredoc contents (no absorbing other lines), the beging is the only doc portion printed in place so trailing commas and method chaining works fine. At this point, _all heredocs are printing themselves correctly anywhere they appear in the AST/Doc. The code used to manually slice up the heredoc and print it correctly in various places (hash.js, array.js, calls.js, utils.js) could now be removed and everything would print valid ruby code.

#### Change in formatting for heredocs in most places

However, now that a lineSuffix with a literalline is appended to the line with the heredoc, the `literalline` seems to break the `group`s and all parent `group`s, resulting in broken-up expressions wherever a heredoc appears. Here's an example:

**Before these changes:**

This expression would stay the same before and after formatting:

```ruby
call(1, <<-TEXT.squish)
  a sentence
TEXT
```

**After these changes:**

```ruby
call(
  1,
  <<-TEXT.squish
  a sentence
TEXT
)
```

I suspect because we're adding a literalline in the linesuffix of a heredoc printed doc, the result is that all parent groups are broken and can no longer be printed on a single line.

My opinion is that this is not undesirable, products consistent heredocs everywhere with more vertical space and potentially less confusion with those less familiar with heredocs. For example, two heredocs on the same line would now print differently:

Input:

```ruby
command 1, 2, 3, <<-HERE, <<-THERE
  here
HERE
  there
THERE
```

Becomes:

```ruby
command 1,
        2,
        3,
        <<-HERE,
  here
HERE
        <<-THERE
  there
THERE
```

jbielick commented on package.json:
```diff
@@ -6,7 +6,7 @@
     "lint": "eslint --cache .",
     "print": "prettier --plugin=.",
-    "test": "jest"
+    "test": "jest --runInBand"
```
will revert this

jbielick commented on src/utils.js:
```diff
@@ -14,6 +8,27 @@ const containsAssignment = (node) =>
   ["assign", "massign"].includes(node.type) ||
   (node.type === "paren" && node.body[0].body.some(containsAssignment));
 
+const isHeredoc = (node) => {
```
removing this as it wasn't needed

jbielick commented on test/js/strings.test.js:
```diff
@@ -280,118 +279,6 @@ describe("strings", () => {
 
         return expect(content).toMatchFormat();
       });
-
```
moved to calls.test.js

jbielick commented on test/js/strings.test.js:
```diff
@@ -219,18 +230,6 @@ describe("strings", () => {
         return expect(content).toMatchFormat();
       });
-
-      test("with a call and indented", () => {
```
moved to calls.test.js

jbielick commented on test/js/strings.test.js:
```diff
@@ -169,6 +169,17 @@ describe("strings", () => {
         return expect(content).toMatchFormat();
       });
 
+      // xstring_literal has no @backtick in the scanner events
```
noticed this, but don't have a fix for it

kddeisz commented:
Thank you very much for your work on this! It certainly does simplify the code a lot.

I agree that it produces more consistent code as a result, but I'm of the opinion that it's less idiomatically Ruby and somewhat less readable in some examples.

Ideally I'd love to have code that produces the following results:

```ruby
my_method(foo, <<~BAR.squish, baz)
  this is my content
BAR

[foo, <<~BAR.squish, baz]
  this is my content
BAR

{
  foo: <<~BAR.squish,
    this is my content
  BAR
  baz: :baz
}
```

At the moment I believe this is working on master. What breaks is if you add an argument to `squish`.

If you'd like to tackle that problem, I'd love to see a PR for that!

Thanks again for your work on this.

jbielick commented:
Hey @kddeisz . I think this branch might be an older iteration. My mistake. When finishing another iteration I intended to reopen a new PR with an updated description. What we ended up with was this (which I think is along the lines of your described desired formatting):

https://github.com/AdWerx/plugin-ruby/pull/2

But if you've already fixed it, that's great. We'll try again to see if it formats our code without produces invalid syntax.

#### Result:

Thus, the end of line break and next few lines are guaranteed to be the heredoc contents (no absorbing other lines), the beging is the only doc portion printed in place so trailing commas and method chaining works fine. At this point, _all heredocs are printing themselves correctly anywhere they appear in the AST/Doc. The code used to manually slice up the heredoc and print it correctly in various places (hash.js, array.js, calls.js, utils.js) could now be removed and everything would print valid ruby code.

All code handling heredocs manually is removed, lines don't always break, but they do if they have to.

Examples:

#### Calls w/ blocks

```js
    test("#301 - heredoc arg w/ block", () => {
      const content = ruby(`
        puts(<<~TEXT
          Hello
        TEXT
        ) { "sample block" }
      `);

      return expect(content).toChangeFormat(
        ruby(`
          puts(<<~TEXT) { 'sample block' }
            Hello
          TEXT
        `)
      );
    });
```

```js
    test("#301 - in an activerecord scope arg w/ chaining", () => {
      const content = ruby(`
        scope :late_for_checkin, -> {
          select(
            <<-EOS.squish
              devices.*, (
                SELECT device_readings.id
                FROM device_readings
                WHERE device_readings.device_id = devices.id
                  AND device_readings.taken_at > DATE_SUB(NOW(), INTERVAL (frequency * 2) MINUTE)
                LIMIT 1
              ) AS reading_id
            EOS
          )
            .data_push
            .having("reading_id IS NULL")
        }
      `);

      return expect(content).toChangeFormat(
        ruby(`
          scope :late_for_checkin,
                -> { select(<<-EOS.squish).data_push.having('reading_id IS NULL') }
                devices.*, (
                  SELECT device_readings.id
                  FROM device_readings
                  WHERE device_readings.device_id = devices.id
                    AND device_readings.taken_at > DATE_SUB(NOW(), INTERVAL (frequency * 2) MINUTE)
                  LIMIT 1
                ) AS reading_id
              EOS
       `)
      );
    });
```

#### In and around args

```js
    test("#586 - long breakable arg after heredoc literal", () => {
      const content = ruby(`
        p <<-BAR, ['value', 'value', 'value', 'value', 'value', 'value', 'value', 'value', 125484, 024024103]
          text
        BAR
      `);

      return expect(content).toChangeFormat(
        ruby(`
          p <<-BAR,
            text
          BAR
            [
              'value',
              'value',
              'value',
              'value',
              'value',
              'value',
              'value',
              'value',
              125_484,
              0o24024103
            ]
        `)
      );
    });
```

```js
    test("#586 - call w/ short breakable arg after heredoc literal", () => {
      const content = ruby(`
        p(<<-BAR, ['value', 'value', 125_484, 0o24024103])
          text
        BAR
      `);

      return expect(content).toMatchFormat();
    });
```

```js
    test("in paren args with a call", () => {
      const content = ruby(`
        Foo.new(<<~ARG1.upcase.chomp, 'test2')
          test1 line 1
          test1 line 2
        ARG1
      `);

      return expect(content).toMatchFormat();
    });
```

```js

    test("on command calls with multiple", () => {
      const content = ruby(`
        command.call 1, 2, 3, <<-HERE, <<-THERE
          here
        HERE
            there
        THERE
      `);

      return expect(content).toMatchFormat();
    });
```

```js
    test("squiggly indent", () => {
      const content = ruby(`
        foo = (<<~TEXT
          bar
        TEXT
        ).strip
      `);

      return expect(content).toChangeFormat(
        ruby(`
          foo = (<<~TEXT).strip
            bar
          TEXT
        `)
      );
    });
```

#### Hashes:

```js
    test("when exceeding line length", () => {
      const content = ruby(`
        { foo: 'foo', bar: <<-HERE, three: 'four', five: 'six', seven: 'eight', nine: 'ten' }
          this is the heredoc
        HERE
      `);

      return expect(content).toChangeFormat(
        ruby(`
          {
            foo: 'foo',
            bar: <<-HERE,
            this is the heredoc
          HERE
            three: 'four',
            five: 'six',
            seven: 'eight',
            nine: 'ten'
          }
        `)
      );
    });
```

#### Arrays

```js
    test("as the last value", () => {
      const content = ruby(`
        [
          foo,
          ${start}
            this is the heredoc
          HERE
        ]
      `);

      return expect(content).toChangeFormat(
        ruby(`
          [foo, ${start}]
              this is the heredoc
            HERE
        `)
      );
    });
```

I found this one particularly interesting:

### Nested heredocs

```js
      test("nested within another", () => {
        const content = ruby(`
          <<~PARENT
            This is a squiggly heredoc
            #{
            <<~CHILD
              This is an interpolated squiggly heredoc
              CHILD
          }
          PARENT
        `);

        return expect(content).toChangeFormat(
          ruby(`
            <<~PARENT
              This is a squiggly heredoc
              #{<<~CHILD}
                This is an interpolated squiggly heredoc
                CHILD
            PARENT
          `)
        );
      });
```

If you prefer this approach, let me know and I can update the PR.

jbielick commented:
@kddeisz that branch / the work there also handles chaining on heredoc literals:

```js
    test("in paren args with a call", () => {
      const content = ruby(`
        Foo.new(<<~ARG1.upcase.chomp, 'test2')
          test1 line 1
          test1 line 2
        ARG1
      `);

      return expect(content).toMatchFormat();
    });
```

jbielick commented:
To be honest, the feedback loop here is not great, so I'll pass on opening another PR. Looks like you've been doing some work merging fixes and addressing issues.

kddeisz commented:
You know what, you're totally right about this. I'm sorry I didn't come around to your way of thinking earlier.

kddeisz commented:
Everything is all kinds of conflicting now, so I'm going to try to open a new PR with your work, just leaving this open until that happens. 

Thanks for your work on this, it's really really good.

kddeisz commented:
Thank you thank you thank you! I apologize it took me a bit to come around. This is really great great work. I used all of this in #681.
